benchmarks/distributed/rpc/ps/launcher.py

```diff
@@ -0,0 +1,372 @@
+        configurations.trainer_count
+    )
+
+    # warmup run the benchmark
+    benchmark_warmup(
+        configurations, trainer_configurations, ps_configurations, model, train_data, parameter_server_rrefs
+    )
+    # run the benchmark
+    trainer_futs = call_trainers(
+        configurations, trainer_configurations, model, train_data, parameter_server_rrefs
+    )
+    # collect metrics and print
+    metrics_printer = ProcessedMetricsPrinter()
+    rank_metrics_list = [fut.wait() for fut in trainer_futs]
+    print_metrics(metrics_printer, rank_metrics_list, "trainer")
+    rank_metrics_list = get_ps_metrics(configurations, ps_configurations, parameter_server_rrefs)
+    print_metrics(metrics_printer, rank_metrics_list, "parameter server")
+
+
+def run_benchmark(rank, model, data, configurations, trainer_configurations, ps_configurations):
+
+    world_size = configurations.world_size
+    os.environ['MASTER_ADDR'] = configurations.master_addr
+    os.environ['MASTER_PORT'] = configurations.master_port
+    rpc_backend_options = TensorPipeRpcBackendOptions()
+    rpc_backend_options.init_method = configurations.rpc_init_method
+
+    # trainers = [0, trainer_count)
+    # parameter_servers = [trainer_count, trainer_count + parameter_server_count)
+    # master = [trainer_count + parameter_server_count, trainer_count + parameter_server_count]
+    if rank == world_size - 1:
+        run_master(rank, model, data, configurations, trainer_configurations, ps_configurations, rpc_backend_options)
+    elif rank >= configurations.trainer_count:
+        rpc.init_rpc(
+            get_name(
+                rank,
+                configurations.trainer_count,
+                configurations.parameter_server_count
+            ),
+            rank=rank,
+            world_size=world_size,
+            rpc_backend_options=rpc_backend_options
+        )
+    else:
+        if (USE_CUDA_RPC in trainer_configurations and
+            trainer_configurations[USE_CUDA_RPC] and
+            USE_CUDA_RPC in ps_configurations and
+                ps_configurations[USE_CUDA_RPC]):
+            ps_rank = get_parameter_server_rank(rank, configurations)
+            ps_name = get_name(
+                ps_rank,
+                configurations.trainer_count,
+                configurations.parameter_server_count
+            )
+            rpc_backend_options.set_device_map(
+                ps_name,
+                {rank: ps_rank}
+            )
+        trainer_name = get_name(
+            rank,
+            configurations.trainer_count,
+            configurations.parameter_server_count
+        )
+        rpc.init_rpc(
+            trainer_name,
+            rank=rank,
+            world_size=world_size,
+            rpc_backend_options=rpc_backend_options
+        )
+    rpc.shutdown()
+
+
+def load_configurations(args):
+    benchmark_config_file = "configurations/benchmark_configurations.json"
+    benchmark_config = json.load(
+        open(
```

Comment:
Don't we need a context manager when opening files? (e.g. `with open(file, "r") as f: ...` ), or does json.load handle that for you?

Reply:
No I use json to parse a file object.
```
f = open(file, "r")
json.load(f)
```

Reply:
Then we should update this to also close the file right? It is always best practice to close after opening to ensure resources (e.g. file descriptor) are released correctly. Your example above would be changed to:
```python
with open(file, "r") as f:
    json.load(f)
```

Reply:
yes forgot to close the file. Ill use close it.